### PR TITLE
Switch input/output to borsh

### PR DESCRIPTION
**Pull request by evgenykuzyakov**
Compiles, but I didn't test the wasm binary yet.

### CODE REVIEW

File: ethtypes/src/lib.rs [hunk 1]
```diff
@@ -117,7 +120,8 @@ pub type Signature = H520;
 
 // Block Header
 
-#[derive(Debug, Clone, Serialize, Deserialize)]
+#[derive(Debug, Clone, BorshDeserialize)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Serialize, Deserialize))]
```
Comment (nearmax) [inline ethtypes/src/lib.rs]: Why `derive(Serialize, Deserialize)` is needed here and in other places?

Comment (evgenykuzyakov) [reply ethtypes/src/lib.rs]: It's used in tests. A lot of test data files are in json

Comment (k06a) [reply ethtypes/src/lib.rs]: I also noticed you moved related deps to dev-deps 👍

**Comment by evgenykuzyakov** [comment]
Got JS borsh ethrelay to work
```
EthBridge is not initialized, initializing...
EthBridge initialization finished
EthBridge check initialization...
EthBridge initialized properly
Contract block number is 0
Web3 block number is 9424991
Computing for block #9424991
Computing for block #9424992
Proofs computation took 8.59s (4.29s per header)
ethBridgeContract.last_block_number = <BN: 0>
Submitting 2 blocks from 9424991 to 9424992 to EthBridge
Blocks submission took 0.76s (0.38s per header)
Successfully submitted 2 blocks from 9424991 to 9424992 to EthBridge
Computing for block #9424992
Computing for block #9424993
Proofs computation took 9.37s (4.68s per header)
ethBridgeContract.last_block_number = <BN: 8fd060>
Submitting 2 blocks from 9424992 to 9424993 to EthBridge
Blocks submission took 0.76s (0.38s per header)
Successfully submitted 2 blocks from 9424992 to 9424993 to EthBridge
Computing for block #9424993
Computing for block #9424994
Proofs computation took 8.26s (4.13s per header)
ethBridgeContract.last_block_number = <BN: 8fd061>
Submitting 2 blocks from 9424993 to 9424994 to EthBridge
Blocks submission took 0.8s (0.4s per header)
Successfully submitted 2 blocks from 9424993 to 9424994 to EthBridge
Computing for block #9424994
Computing for block #9424995
Computing for block #9424996
Computing for block #9424997
Proofs computation took 16.39s (4.09s per header)
ethBridgeContract.last_block_number = <BN: 8fd062>
Submitting 4 blocks from 9424994 to 9424997 to EthBridge
Blocks submission took 1.23s (0.3s per header)
Successfully submitted 4 blocks from 9424994 to 9424997 to EthBridge
Computing for block #9424997
Computing for block #9424998
Proofs computation took 8.4s (4.2s per header)
ethBridgeContract.last_block_number = <BN: 8fd065>
Submitting 2 blocks from 9424997 to 9424998 to EthBridge
Blocks submission took 0.75s (0.37s per header)
Successfully submitted 2 blocks from 9424997 to 9424998 to EthBridge
Computing for block #9424998
Computing for block #9424999
Computing for block #9425000
Proofs computation took 12.66s (4.22s per header)
ethBridgeContract.last_block_number = <BN: 8fd066>
Submitting 3 blocks from 9424998 to 9425000 to EthBridge
Blocks submission took 0.96s (0.32s per header)
Successfully submitted 3 blocks from 9424998 to 9425000 to EthBridge
```

**Comment by evgenykuzyakov** [comment]
At the end I hit:
```
Blocks submission took 0.96s (0.32s per header)
Successfully submitted 3 blocks from 9425020 to 9425022 to EthBridge
Computing for block #9425022
Computing for block #9425023
Computing for block #9425024
Computing for block #9425025
Computing for block #9425026
Proofs computation took 22.05s (4.41s per header)
ethBridgeContract.last_block_number = <BN: 8fd07e>
Submitting 5 blocks from 9425022 to 9425026 to EthBridge
(node:92271) UnhandledPromiseRejectionWarning: Error: The used access key requires exactly one FunctionCall action
    at Object.parseRpcError (/Users/ekwork/code/near-bridge/ethrelay/node_modules/nearlib/lib/utils/rpc_errors.js:25:19)
    at Account.signAndSendTransaction (/Users/ekwork/code/near-bridge/ethrelay/node_modules/nearlib/lib/account.js:87:36)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async EthBridgeContract.value [as add_block_headers] (/Users/ekwork/code/near-bridge/ethrelay/index.js:253:39)
    at async /Users/ekwork/code/near-bridge/ethrelay/index.js:406:9
    at async Subscription.callback (/Users/ekwork/code/near-bridge/ethrelay/index.js:33:13)
(node:92271) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:92271) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

```
Probably gas limit exceeded

**Comment by evgenykuzyakov** [comment]
Looks like I got it to work consistently with uploading 2 blocks at a time on top of 1 old block:
```
locks submission took 1.03s (0.34s per header)
Successfully submitted 3 blocks from 9425126 to 9425127 to EthBridge
Computing for block #9425130 to #9425132
ethBridgeContract.last_block_number = <BN: 8fd0e7>
Submitting 3 blocks from 9425128 to 9425129 to EthBridge
TxHash 769aa6hQT3FAaZ42a668CaVfZtB1waYQYJei9j5JFkRZ
Blocks submission took 0.99s (0.33s per header)
Successfully submitted 3 blocks from 9425128 to 9425129 to EthBridge
Computing for block #9425132 to #9425134
ethBridgeContract.last_block_number = <BN: 8fd0e9>
Submitting 3 blocks from 9425130 to 9425131 to EthBridge
TxHash 4NbxcKqK815fTy7g5G9eZb9Pove2uuuzwFDtjWNy83Fi
Blocks submission took 0.98s (0.32s per header)
Successfully submitted 3 blocks from 9425130 to 9425131 to EthBridge
Computing for block #9425134 to #9425136
ethBridgeContract.last_block_number = <BN: 8fd0eb>
Submitting 3 blocks from 9425132 to 9425133 to EthBridge
TxHash 8A3QgB9sXB3X1bD24ydJJSnzpGMUsxpZ44MrxYKtknU2
Blocks submission took 0.99s (0.33s per header)
Successfully submitted 3 blocks from 9425132 to 9425133 to EthBridge
Computing for block #9425136 to #9425138
ethBridgeContract.last_block_number = <BN: 8fd0ed>
Submitting 3 blocks from 9425134 to 9425135 to EthBridge
TxHash 9eqaAwzfgz81FzimMdDsAeVd6fPnRvGoRavoT6c7VCxQ
Blocks submission took 1.17s (0.39s per header)
Successfully submitted 3 blocks from 9425134 to 9425135 to EthBridge
Computing for block #9425138 to #9425140
ethBridgeContract.last_block_number = <BN: 8fd0ef>
Submitting 3 blocks from 9425136 to 9425137 to EthBridge
TxHash 91MmuiWkm7LbD7PNj8iRAPqoJY3ZW2JHNmTjQoAM58X3
Blocks submission took 1.03s (0.34s per header)
Successfully submitted 3 blocks from 9425136 to 9425137 to EthBridge
Computing for block #9425140 to #9425142
ethBridgeContract.last_block_number = <BN: 8fd0f1>
Submitting 3 blocks from 9425138 to 9425139 to EthBridge
TxHash 5TDp57ZCBxBnaoGiViKXWwfM99cAopvPc88KsMiPHYc2
Blocks submission took 1.03s (0.34s per header)
Successfully submitted 3 blocks from 9425138 to 9425139 to EthBridge
Computing for block #9425142 to #9425144
ethBridgeContract.last_block_number = <BN: 8fd0f3>
Submitting 3 blocks from 9425140 to 9425141 to EthBridge
TxHash DZJbhsoA2cCNmKgFj8PqaC9REsmUR9zj3Q4Rp5p4ymX2
Blocks submission took 0.99s (0.33s per header)
Successfully submitted 3 blocks from 9425140 to 9425141 to EthBridge
Computing for block #9425144 to #9425146
ethBridgeContract.last_block_number = <BN: 8fd0f5>
Submitting 3 blocks from 9425142 to 9425143 to EthBridge
TxHash DU7DgMebg9devcdZxrp6K3yWLTt7MomnfYrqJi3bLCqx
Blocks submission took 1.02s (0.34s per header)
Successfully submitted 3 blocks from 9425142 to 9425143 to EthBridge
Computing for block #9425146 to #9425148
ethBridgeContract.last_block_number = <BN: 8fd0f7>
Submitting 3 blocks from 9425144 to 9425145 to EthBridge
TxHash 3QfqQKao92bTNLrfAPZNcj8Q67G4RCfCBP18FnXHs2oT
Blocks submission took 1.04s (0.34s per header)
Successfully submitted 3 blocks from 9425144 to 9425145 to EthBridge
Computing for block #9425148 to #9425150
ethBridgeContract.last_block_number = <BN: 8fd0f9>
Submitting 3 blocks from 9425146 to 9425147 to EthBridge
TxHash 4BXgsVTPMVFd4Dd8ViYhp3wuP5gNJPGazsfapoQyBWDt
Blocks submission took 1.08s (0.36s per header)
Successfully submitted 3 blocks from 9425146 to 9425147 to EthBridge
Computing for block #9425150 to #9425152
ethBridgeContract.last_block_number = <BN: 8fd0fb>
Submitting 3 blocks from 9425148 to 9425149 to EthBridge
TxHash 52Uhy5f4ziVyfXTwk8odGLyvMDxrS5FJFa18AbYtEj8d
Blocks submission took 1.05s (0.35s per header)
Successfully submitted 3 blocks from 9425148 to 9425149 to EthBridge
Computing for block #9425152 to #9425154
ethBridgeContract.last_block_number = <BN: 8fd0fd>
Submitting 3 blocks from 9425150 to 9425151 to EthBridge
TxHash 9Um7vYZqibXxsMSsRfREshEgjCXirJz8XPdwHVm9GFXo
Blocks submission took 1s (0.33s per header)
Successfully submitted 3 blocks from 9425150 to 9425151 to EthBridge
Computing for block #9425154 to #9425156
ethBridgeContract.last_block_number = <BN: 8fd0ff>
Submitting 3 blocks from 9425152 to 9425153 to EthBridge
TxHash DP3Na5oZXbK9Dwo8hjZJ6G2o99c1ehfyWcLRwSWSJShB
Blocks submission took 0.99s (0.33s per header)
Successfully submitted 3 blocks from 9425152 to 9425153 to EthBridge
Computing for block #9425156 to #9425158
ethBridgeContract.last_block_number = <BN: 8fd101>
Submitting 3 blocks from 9425154 to 9425155 to EthBridge
TxHash TuPZc7ic38o7QsVD8xJZYZGwuVg4pt2CUkrVAaXvbPf
Blocks submission took 1.06s (0.35s per header)
Successfully submitted 3 blocks from 9425154 to 9425155 to EthBridge
Computing for block #9425158 to #9425160
ethBridgeContract.last_block_number = <BN: 8fd103>
Submitting 3 blocks from 9425156 to 9425157 to EthBridge
TxHash GmW9o4AMGSvEh7mN2C3uUCCUMnSR3YVsUMFSGi6gHxEY
Blocks submission took 1.41s (0.47s per header)
Successfully submitted 3 blocks from 9425156 to 9425157 to EthBridge
Computing for block #9425160 to #9425162
ethBridgeContract.last_block_number = <BN: 8fd105>
Submitting 3 blocks from 9425158 to 9425159 to EthBridge
TxHash Fp7KZXtk6vPCwAvdM79iSzafehpTfRwTVvzfT4YgPES5
Blocks submission took 1.08s (0.36s per header)
Successfully submitted 3 blocks from 9425158 to 9425159 to EthBridge
Computing for block #9425162 to #9425164
ethBridgeContract.last_block_number = <BN: 8fd107>
Submitting 3 blocks from 9425160 to 9425161 to EthBridge
TxHash FX89G76w5FXDHam7d8KTvwV1FYW4qo66gTnTPtVqXiB7
Blocks submission took 1.06s (0.35s per header)
Successfully submitted 3 blocks from 9425160 to 9425161 to EthBridge
Computing for block #9425164 to #9425166
ethBridgeContract.last_block_number = <BN: 8fd109>
Submitting 3 blocks from 9425162 to 9425163 to EthBridge
TxHash Fz2pnBg2wCcyP2T91EmBhbKECkuUCUPFT8pgijJVioZD
Blocks submission took 0.99s (0.33s per header)
Successfully submitted 3 blocks from 9425162 to 9425163 to EthBridge
```

File: ethbridge/src/lib.rs [hunk 2]
```diff
@@ -190,10 +194,10 @@ impl EthBridge {
 
     pub fn hashimoto_merkle(
         &self,
-        header_hash: H256,
-        nonce: H64,
+        header_hash: &H256,
+        nonce: &H64,
         block_number: u64,
-        nodes: Vec<DoubleNodeWithMerkleProof>,
```
Comment (k06a) [inline ethbridge/src/lib.rs]: I see potential problem to switch back to `near-bindgen`, when arguments parsing will use Borsh  by default

Comment (nearmax) [reply ethbridge/src/lib.rs]: Yes, let's not switch to auto-borsh parsing until after EthDenver.

Comment (evgenykuzyakov) [reply ethbridge/src/lib.rs]: I split `impl EthBridge {` into 2 sections. One to use with `near-bindgen` macro and one to use without. Basically, they are private methods that only used internally and don't expose function API publicly from contract.

**Pull request merged**